README.md
# 下班提醒插件（go-off-work）

这是你的扩展 "下班提醒插件（go-off-work）" 的 README 文件。

## 功能

该插件可以在状态栏中显示距离下班还有多少时间，并在下班时间到达时提醒你。

## 需求

无特殊需求或依赖项。

## 扩展设置

该扩展支持以下设置：

- `goOffWork.hour`：设置下班的小时数。
- `goOffWork.minute`：设置下班的分钟数。
- `goOffWork.notificationMessage`设置右下角通知信息
- `goOffWork.getOffMessage`设置到时间提示信息

你可以通过修改这些设置来自定义下班时间。

## 使用方法

在 Visual Studio Code 中，你可以按下 `Ctrl+Shift+P`（或 `Cmd+Shift+P`）来打开命令面板，然后输入 "设置下班时间" 并选择它，即可修改下班时间。

## 发布说明

### 0.0.1

初始发布，提供基本的下班提醒功能。

### 0.0.2

修复扩展加载后不显示时间问题

### 0.0.3

支持命令  `设置下班时间` 时间函数补零

### 0.0.4

新增插件 logo 用的是 崩坏星穹铁道 青雀角色头像

### 0.0.5

新增 单元测试，优化 不停创建 statusBar 导致的闪屏问题

### 0.0.6

新增 支持通知自定义 `notificationMessage` 消息自定义  `getOffMessage`

### 0.0.7

修复了一下，输入的时间限制 小时只能 `00 - 23` 分钟只能 `00 - 59`

---

## 遵循扩展准则

确保你已阅读扩展准则，并遵循最佳实践来创建你的扩展。

- [扩展准则](https://code.visualstudio.com/api/references/extension-guidelines)

**尽情享受！**


CHANGELOG.md
# 版本更新记录

"下班提醒插件（go-off-work）" 扩展的重要更新将在此文件中记录。

## [0.0.7] - 2023-08-24

* 修复了一下，输入的时间限制 小时只能 `00 - 23` 分钟只能 `00 - 59`

## [0.0.6] - 2023-08-24

* 新增 支持通知自定义 `notificationMessage` 消息自定义  `getOffMessage`

## [0.0.5] - 2023-08-22

* 新增 单元测试，优化 不停创建 statusBar 导致的闪屏问题

## [0.0.4] - 2023-08-21

* 新增插件 logo 用的是 崩坏星穹铁道 青雀角色头像

## [0.0.3] - 2023-08-21

* 支持命令  `设置下班时间` 时间函数补零

## [0.0.2] - 2023-08-21

* 修复扩展加载后不显示时间问题

## [0.0.1] - 2023-08-21

* 初始发布，提供基本的下班提醒功能。


test/index.test.ts
import { describe, it, expect } from "vitest";
import { addLeadingZero, messageTemplate, getMessage } from "../src/utils";

describe("utils functions", () => {
  it("addLeadingZero 函数", () => {
    expect(addLeadingZero(0)).toBe("00");
    expect(addLeadingZero(1)).toBe("01");
    expect(addLeadingZero(2)).toBe("02");
    expect(addLeadingZero(12)).toBe("12");
    expect(addLeadingZero(13)).toBe("13");
  });

  it("messageTemplate 函数", () => {
    expect(messageTemplate(0, 0, 0)).toBe(">> 距离下班还有 00小时00分钟00秒");
    expect(messageTemplate(1, 0, 0)).toBe(">> 距离下班还有 01小时00分钟00秒");
    expect(messageTemplate(2, 0, 0)).toBe(">> 距离下班还有 02小时00分钟00秒");
    expect(messageTemplate(12, 0, 0)).toBe(">> 距离下班还有 12小时00分钟00秒");
    expect(messageTemplate(13, 0, 0)).toBe(">> 距离下班还有 13小时00分钟00秒");

    expect(messageTemplate(0, 1, 0)).toBe(">> 距离下班还有 00小时01分钟00秒");
    expect(messageTemplate(0, 2, 0)).toBe(">> 距离下班还有 00小时02分钟00秒");
    expect(messageTemplate(0, 30, 0)).toBe(">> 距离下班还有 00小时30分钟00秒");
    expect(messageTemplate(0, 59, 0)).toBe(">> 距离下班还有 00小时59分钟00秒");

    expect(messageTemplate(0, 0, 1)).toBe(">> 距离下班还有 00小时00分钟01秒");
    expect(messageTemplate(0, 0, 2)).toBe(">> 距离下班还有 00小时00分钟02秒");
    expect(messageTemplate(0, 0, 30)).toBe(">> 距离下班还有 00小时00分钟30秒");
    expect(messageTemplate(0, 0, 59)).toBe(">> 距离下班还有 00小时00分钟59秒");

    expect(messageTemplate(1, 30, 45)).toBe(">> 距离下班还有 01小时30分钟45秒");
    expect(messageTemplate(5, 15, 30)).toBe(">> 距离下班还有 05小时15分钟30秒");
    expect(messageTemplate(10, 45, 20)).toBe(
      ">> 距离下班还有 10小时45分钟20秒"
    );
    expect(messageTemplate(23, 59, 59)).toBe(
      ">> 距离下班还有 23小时59分钟59秒"
    );
  });

  it("getMessage 函数", () => {
    const now = new Date(); // 获取当前日期
    const fixedDateTime = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 10, 0);

    expect(getMessage(fixedDateTime, { hour: 18, minute: 0 })).toBe(
      ">> 距离下班还有 08小时00分钟00秒"
    );

    expect(getMessage(fixedDateTime, { hour: 18, minute: 10 })).toBe(
      ">> 距离下班还有 08小时10分钟00秒"
    );

    expect(getMessage(fixedDateTime, { hour: 18, minute: 20 })).toBe(
      ">> 距离下班还有 08小时20分钟00秒"
    );

    expect(getMessage(fixedDateTime, { hour: 17, minute: 0 })).toBe(
      ">> 距离下班还有 07小时00分钟00秒"
    );

    expect(getMessage(fixedDateTime, { hour: 10, minute: 0, getOffMessage: "xxxxxxxxxxx" })).toBe(
      "xxxxxxxxxxx"
    );
  });
});


src/utils/index.ts
import { getOffMessage } from "../constant";
import type { Config } from "../types";

// 补零函数，确保时间显示两位数
export const addLeadingZero = (num: number) => num.toString().padStart(2, "0");

// 消息模版
export const messageTemplate = (
  hour: number,
  minute: number,
  second: number
): string => {
  return `>> 距离下班还有 ${addLeadingZero(hour)}小时${addLeadingZero(
    minute
  )}分钟${addLeadingZero(second)}秒`;
};

// 获取提示消息
export function getMessage(now: Date, config: Config): string {
  const goOffWork = new Date();
  goOffWork.setHours(config.hour, config.minute, 0, 0); // 将秒和毫秒设置为0

  const duration = Math.max(goOffWork.getTime() - now.getTime(), 0);

  if (duration <= 0) {
    return config.getOffMessage || getOffMessage;
  }

  const hour = Math.floor(duration / 1000 / 60 / 60);
  const minute = Math.floor((duration / 1000 / 60) % 60);
  const second = Math.floor((duration / 1000) % 60);

  return messageTemplate(hour, minute, second);
}



src/extension.ts
import * as vscode from "vscode";
import type { Config } from ".//types";
import { getMessage } from "./utils";
import { getOffMessage, notificationMessage } from "./constant";

let timer: NodeJS.Timeout | undefined;
let statusBar: vscode.StatusBarItem | undefined;

// 创建并显示状态栏消息
function showStatusBarMessage(message: string) {
  if (!statusBar) {
    statusBar = vscode.window.createStatusBarItem(
      vscode.StatusBarAlignment.Left,
      0
    );
    statusBar.show();
  }
  statusBar.text = message;
}

// 更新状态栏项
function updateStatusBarItem(config: Config) {
  const newMessage = getMessage(new Date(), config);
  showStatusBarMessage(newMessage);

  if (newMessage === (config.getOffMessage || getOffMessage)) {
    vscode.window.showInformationMessage(
      config.notificationMessage || notificationMessage
    );
    clearInterval(timer!);
  } else {
    timer = setTimeout(() => {
      updateStatusBarItem(config);
    }, 1000);
  }
}

// 设置下班时间的命令
async function setGoOffWorkTime() {
  const config = vscode.workspace
    .getConfiguration()
    .get<Config>("goOffWork") as Config;

  const inputValue = await vscode.window.showInputBox({
    placeHolder: "请输入下班时间（HH:mm）",
  });

  if (inputValue) {
    const [hour, minute] = inputValue.split(":");
    const parsedHour = parseInt(hour, 10);
    const parsedMinute = parseInt(minute, 10);

    if (
      !isNaN(parsedHour) &&
      !isNaN(parsedMinute) &&
      parsedHour >= 0 &&
      parsedHour < 24 &&
      parsedMinute >= 0 &&
      parsedMinute < 60
    ) {
      config.hour = parsedHour;
      config.minute = parsedMinute;

      const configuration = vscode.workspace.getConfiguration();
      await configuration.update(
        "goOffWork.hour",
        config.hour,
        vscode.ConfigurationTarget.Global
      );
      await configuration.update(
        "goOffWork.minute",
        config.minute,
        vscode.ConfigurationTarget.Global
      );

      // 通知
      vscode.window.showInformationMessage(`下班时间设置成功 >> ${inputValue}`);
    } else {
      vscode.window.showErrorMessage(
        `请输入有效的时间格式 (HH:mm) >> ${inputValue} ???`
      );
    }
  }
}

// 设置通知信息 `notificationMessage`
async function setNotificationMessage() {
  const config = vscode.workspace
    .getConfiguration()
    .get<Config>("goOffWork") as Config;

  const inputValue = await vscode.window.showInputBox({
    placeHolder: "请输入通知信息",
  });

  if (inputValue) {
    config.notificationMessage = inputValue;

    const configuration = vscode.workspace.getConfiguration();
    await configuration.update(
      "goOffWork.notificationMessage",
      config.notificationMessage,
      vscode.ConfigurationTarget.Global
    );
  }
}

// 设置提示信息 `getOffMessage`
async function setGetOffMessage() {
  const config = vscode.workspace
    .getConfiguration()
    .get<Config>("goOffWork") as Config;

  const inputValue = await vscode.window.showInputBox({
    placeHolder: "请输入提示信息",
  });

  if (inputValue) {
    config.getOffMessage = inputValue;

    const configuration = vscode.workspace.getConfiguration();
    await configuration.update(
      "goOffWork.getOffMessage",
      config.getOffMessage,
      vscode.ConfigurationTarget.Global
    );
  }
}

export function activate(context: vscode.ExtensionContext) {
  const allConfig = vscode.workspace.getConfiguration();
  let config: Config = allConfig.goOffWork;

  if (
    !config ||
    typeof config.hour !== "number" ||
    typeof config.minute !== "number"
  ) {
    vscode.window.showErrorMessage("请正确配置 goOffWork 的 hour 和 minute！");
    return;
  }

  updateStatusBarItem(config);

  // 监听配置变化，在配置变化时重新加载插件
  vscode.workspace.onDidChangeConfiguration((event) => {
    if (event.affectsConfiguration("goOffWork")) {
      config = vscode.workspace
        .getConfiguration()
        .get<Config>("goOffWork") as Config;
      clearInterval(timer!);
      updateStatusBarItem(config);
    }
  });

  const commandMappings = [
    // 注册设置下班时间的命令
    { command: "setGoOffWorkTime", handler: setGoOffWorkTime },
    // 注册设置通知信息的命令
    { command: "setNotificationMessage", handler: setNotificationMessage },
    // 注册设置提示信息的命令
    { command: "setGetOffMessage", handler: setGetOffMessage },
  ];

  commandMappings.forEach(({ command, handler }) => {
    // 注册命令
    context.subscriptions.push(
      vscode.commands.registerCommand(command, handler)
    );
  });
}

export function deactivate() {
  if (timer) {
    clearTimeout(timer);
  }
  if (statusBar) {
    statusBar.dispose();
  }
}
